qiime/core/type/signature.py

```diff
@@ -9,18 +9,74 @@
 import collections
 import inspect
 import itertools
+import copy
 
 import qiime.sdk
 from .grammar import TypeExpression
 from .primitive import is_primitive_type
 from .semantic import is_semantic_type
 from .visualization import Visualization
+from ..util import ImmutableBase
+
+
+class _NoValue:
+    def __repr__(self):
+        return "NOVALUE"
+
+
+class ParameterSpec(ImmutableBase):
+    def __init__(self, **kwargs):
+        self.NOVALUE = _NoValue()
+        self.qiime_type = kwargs.pop('qiime_type', self.NOVALUE)
+        self.view_type = kwargs.pop('view_type', self.NOVALUE)
+        self.default = kwargs.pop('default', self.NOVALUE)
+        if kwargs:
+            raise TypeError("Got unexpected keyword argument(s) %r"
+                            % list(kwargs))
+        self._freeze_()
+
+    def has_qiime_type(self):
+        return self.qiime_type is not self.NOVALUE
+
+    def has_view_type(self):
+        return self.view_type is not self.NOVALUE
+
+    def has_default(self):
+        return self.default is not self.NOVALUE
+
+    def __repr__(self):
+        return ("ParameterSpec(qiime_type=%r, view_type=%r, default=%r)"
+                % (self.qiime_type, self.view_type, self.default))
+
+    def __eq__(self, other):
+        if self.has_qiime_type() != other.has_qiime_type():
+            return False
+        if self.has_qiime_type() and not self.qiime_type.equals(
```

Comment:
Can this `__eq__` be simplified to so that `==` is used to compare each property? I think this is possible once a module-level sentinel is used, but I'm not sure what the difference is between `__eq__` and `equals` on the QIIME types.

```python
return (
    self.qiime_type == other.qiime_type and
    self.view_type == other.view_type and
    self.default == other.default
)
```